### DISCUSSION TITLE
Licensing of a plugin reusing parts of pretix' code

### DISCUSSION THREAD

**Discussion by bockstaller** [pretix/pretix#2286] [Q&A] [answered]
I am currently writing a [plugin](https://github.com/bockstaller/pretix-batch-emailer) to send emails to groups of orders and would like to cherrypick and modify code from the [sendmail](https://github.com/pretix/pretix/tree/master/src/pretix/plugins/sendmail) plugin (with the option to upstream the changes, but I want to get it out of the door as fast as possible).

Now I am confused by your [Licensing FAQ](https://docs.pretix.eu/en/latest/license/faq.html) and AGPL in General 😅 

[What licenses can plugins use?](https://docs.pretix.eu/en/latest/license/faq.html#what-licenses-can-plugins-use) hints that I should use the Apache License 2.0 to avoid problems.
[What if I want to re-use a minor part of pretix in my project](https://docs.pretix.eu/en/latest/license/faq.html#what-if-i-want-to-re-use-a-minor-part-of-pretix-in-my-project) says that I would have to license as AGPL (given that it is probably written with projects that aren't part of the pretix universe in mind). 

Can I ignore these mixed messages, use the Apache License 2.0 and be fine? 
Have I missed some licensing clauses completely?

**Comment by raphaelm** [comment, marked as answer]
This is the big pain point in the license model, yes.

(First of all, I am not a lawyer and this is not legal advice.)

If you take code from pretix and want to re-use it in a third-party project, regardless of whether it is a plugin to pretix or not, you **must** comply with AGPLv3 and therefore your plugin **must** be AGPLv3.

However (and that's why the FAQ generally recommends against doing so) if you license your plugin with AGPLv3, it is no longer legally possible to install it on a pretix instance that has *any* non-AGPLv3 code, such as internal custom plugins covered by pretix' additional permissions, or pretix Enterprise plugins. So it would be impossible to install your plugin and pretix Enterprise plugins or non-public custom code on the same instance. (There's *probably* some way around the second part if you don't make your plugin pure-AGPL but use the same extended license terms as the pretix itself, but there's no way around the first part.)

As we say in the FAQ, we're open to the idea of releasing parts of the codebase independently if that helps and does not threaten the integrity of the project itself. However in your specific case there might be a simpler solution… What if you just take Apache-licensed code from pretix 3.17 or older? ;) I don't think the relevant areas changed much since then.

**Reply by bockstaller** [reply]
Ah, that makes the problem clearer. The pretix 3.17 solution should work for me. Thank you!
